### PR TITLE
Data Science V2

### PR DESCRIPTION
**Pull request by mcdaniel67**

### CODE REVIEW

File: data_science/calculator.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+require_relative('json_parser')
+require('abanalyzer')
+
+class Calculator
+  attr_reader :data
+
+  def load(filepath)
+    @data = JSONParser.new.parse(filepath)
```
Comment (zsyed1) [inline data_science/calculator.rb]: You really don't need to store the entire data set. All that matters is the number of conversion and non-conversions (or equivalently the total visits). Imagine you had 10M data points :smile:

File: data_science/calculator.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  def chi_square_confidence
+    values = { a: cohort_totals('A'), b: cohort_totals('B') }
+    ab_test = ABAnalyzer::ABTest.new(values)
+    ab_test.chisquare_p
+  end
+
+  def cohort_totals(cohort)
+    { success: number_of_conversions(cohort), total: total_size(cohort) }
+  end
+end
```
Comment (zsyed1) [inline data_science/calculator.rb]: Your code is concisely written!

File: data_science/calculator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  let(:interval_a) { stat_calc.confidence_interval('A', 0.95) }
+
+  it 'should correctly load a small json dataset' do
+    expect(calculator.data).to eq([
+      { 'date' => '2014-03-20', 'cohort' => 'A', 'result' => 1 },
+      { 'date' => '2014-03-20', 'cohort' => 'A', 'result' => 1 },
+      { 'date' => '2014-03-20', 'cohort' => 'A', 'result' => 0 },
+      { 'date' => '2014-03-20', 'cohort' => 'A', 'result' => 0 },
+    ])
```
Comment (zsyed1) [inline data_science/calculator_spec.rb]: There may be a better way to do this - but yours is definitely readable.

File: data_science/calculator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+require('abanalyzer')
+
+describe Calculator do
+  def build_calculator(filename)
+    calculator = Calculator.new
+    calculator.load(filename)
+    calculator
+  end
+
+  let(:float_tol) { 1e-8 }
```
Comment (zsyed1) [inline data_science/calculator_spec.rb]: The Ruby community prefers full names over abbreviations - maybe `tolerance`.

File: data_science/json_parser_spec.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+require_relative('json_parser.rb')
+
+describe JSONParser do
+  let(:parser) { JSONParser.new }
+
+  it 'should fail elegantly when the file does not exist' do
```
Comment (zsyed1) [inline data_science/json_parser_spec.rb]: I think you meant to say `gracefully`

File: data_science/json_test_files/statistics_test.json [hunk 6]
```diff
@@ -0,0 +1 @@
...:1},{"date":"2014-03-20","cohort":"A","result":0},{"date":"2014-03-20","cohort":"A","result":0},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1},{"date":"2014-03-20","cohort":"B","result":1}]
```
Comment (zsyed1) [inline data_science/json_test_files/statistics_test.json]: Very nice work!

Comment (zsyed1) [inline data_science/json_test_files/statistics_test.json]: Your code is concise, no-fuss, I enjoyed reading it.

**Comment by paulghaddad** [comment]
@mcdaniel67 Nice work on this resubmission. You're good on this exercise.

**Pull request closed without merging**